doc/manpages/ccndcontrol.1
'\" t
.\"     Title: ccndcontrol
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 01/10/2014
.\"    Manual: \ \&
.\"    Source: \ \& 0.8.2
.\"  Language: English
.\"
.TH "CCNDCONTROL" "1" "01/10/2014" "\ \& 0\&.8\&.2" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ccndcontrol \- manipulate CCNx forwarding table
.SH "SYNOPSIS"
.sp
\fBccndcontrol\fR [\-v] \-f \fIconfigfile\fR
.sp
\fBccndcontrol\fR [\-v] add \fIuri\fR (upd|tcp) \fIhost\fR [\fIport\fR [flags [mcastttl [mcastif]]]]
.sp
\fBccndcontrol\fR [\-v] del \fIuri\fR (upd|tcp) \fIhost\fR [\fIport\fR [flags [mcastttl [mcastif]]]]
.SH "DESCRIPTION"
.sp
The \fBccndcontrol\fR utility was written as a Java\-based workalike for ccndc(1), but is not being actively maintained or extended\&. In its current form, it serves as an example of how to use the Java API for prefix registration\&.
.sp
The significance of the operands is as follows:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}

\fB\fIuri\fR\fR
is the name prefix to be registered\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}

\fB\fIhost\fR\fR
identifies the remote hostname for this face\&.
.RE
.sp
See the ccndc(1) man page for a more complete description of the operands\&.
.SH "OPTIONS"
.PP
\fB\-v\fR
.RS 4
Sets the logging level to INFO and prints messages when prefixes are registered and when ccnd faces are created or deleted\&.
.RE
.SH "EXIT STATUS"
.PP
\fB0\fR
.RS 4
Success
.RE
.PP
\fB1\fR
.RS 4
Failure
.RE
.SH "AUTHOR"
.sp
Tim Diebert


NEWS
Changes between 0.8.1 and 0.8.2:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/21 for more details on CCNx 0.8.2.

- This release features a refactored strategy layer, making it straightforward to add new
  strategies in a modular way.  Alongside the traditional "default" strategy there are
  two new examples, a "parallel" strategy that sends interests on all available faces
  and a "loadsharing" strategy that sends each arriving interest to just one of the
  available faces.  Strategies may be selected by prefix.  This selection may be
  done using new subcommands to the ccndc utility.

	* Feature #100929: Implement Junit test suites to facilitate bulk testing.
	* Feature #100983: strategy layer refactoring
	* Feature #100984: strategy labels on prefixes
	* Feature #100990: Restore SYNC_NOISY to INFO
	* Feature #100992: Add test-security to standard java tests
	* Feature #100993: Add test-versioning back into standard java tests
	* Feature #100994: Remove unused cruft from build.xml
	* Feature #101036: java sync should use Deque operations instead of Stack as recommended in the javadocs
	* Feature #101039: UDP mode for java netmanagers is not tested in standard testing
	* Feature #101053: ccncatchunks2 could use improved hole-filling and interest lifetime management
	* Bug #101030: build.xml setup of ccnr environment is incorrect
	* Bug #101033: Handling of extremely large values of MaxSuffixComponents is incorrect
	* Bug #101037: ccn_seqwriter library code should not limit the block size to 4096 bytes
	* Bug #101042: CcnFileProxy uses deprecated methods
	* Bug #101046: VLC 2.1.0 release necessitates change in plugin Makefile
	* Bug #101048: Do not rely on accept(2) returned addr for recognizing a unix-domain socket
	* Bug #101052: Unbreak objtree build target
	* Bug #101057: ccn_sign_content / ccn_chk_signing_params do not properly incorporate a timestamp in a template via ccn_signing_params
	* Bug #101058: ccndc in daemon mode does not exit cleanly when ccnd goes away
	* Bug #101059: testTimeouts a little too strict
	* Bug #101060: ccnd cache lookups should take a controllable amount of work
	* Bug #101061: ccnd minimum interest lifetime problem
	* Bug #101062: some ccnd logging missing check for logging enabled before calling expensive log procedure
	* Support #101047: Prepare for ccnx-0.8.2 release

Changes between 0.8.0 and 0.8.1:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/20 for more details on CCNx 0.8.1.

- This release fixes a number of bugs, including a serious memory leak in ccnd.

	* Bug #101029: unbounded test log file growth
	* Bug #101032: ccnr should start at specified segment on checked start write
	* Bug #101034: No obvious documentation for how to start ccnr exists on website
	* Bug #101035: small leak in ccn_fetch should a realloc() fail
	* Bug #101038: Inconsistent encoding of zero-length name components
	* Bug #101043: ccnr r_store_lookup_backwards can fail
	* Bug #101044: valgrind reports memory leak in 0.8.0 ccnd
	* Bug #101049: ccnd can remove newly arrived content too quickly
	* Support #101045 Prepare for ccnx-0.8.1 release

Changes between 0.7.2 and 0.8.0:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/17 for more details on CCNx 0.8.0.

- The use of symmetric key HMACs in place of public/private key signatures
  is now an option in both libraries.  In situations where it can be used,
  it provides a much higher performance alternative.
- The content store in ccnd has been improved.  It now uses a flatname
  representation for indexing, in the same way that ccnr does.  Also, the
  underlying skiplist implementation has moved into the library, so that
  applications may take advantage of it in the future.  In the process,
  some old ccnd bugs relating to the handling of extreme values of
  FreshnessSeconds were fixed.
- The more readable escaping convention in URIs is now the default.
  If for whatever reason you prefer to use only percent-escaping in URIs, use

	CFLAGS='-g -DCCN_URI_DEFAULT_ESCAPE=4' ./configure

  before typing make.  Note that both forms of escaping are accepted as input.
- An example rc startup script is now included, but needs to be installed by
  hand.  It is coded for Ubuntu Linux, but should be easily adaptable to other
  systems.
- This release drops support for Java JDK 1.5.

  * Issues Resolved:
	* Bug #100118: ccnd does not handle FreshnessSeconds=0 correctly.
	* Bug #100119: ccnd should handle large values of FreshnessSeconds
	* Bug #100558: test-security does not pass
	* Bug #100897: timeout values  should be consistently long throughout java code 
	* Bug #100966: Incorrect processing of CCNS_SYNC_SCOPE=3
	* Bug #100977: Problem with verify in CCNInputStream when trying to retrieve key
	* Bug #100987: Non-portable use of local keyword in Bourne shell scripts
	* Bug #100998: VLC 2.1.x block_New definition removed
	* Bug #100999: java BLOCK_AFTER_FIRST_SEGMENT doesn't work in all cases
	* Bug #101000: sync_diff code should reset differencing data when aborting or completing a comparison
	* Bug #101004: QueuedContentHandler has a sync error
	* Bug #101005: CCNx Android Services build out of sync with API changes to CCNx C-lib
	* Bug #101007: ccn_resolve_version() should distinguish between no response and finding an unversioned object.
	* Bug #101010: ccnseqwriter command busy-loops when its ccnd goes away
	* Bug #101011: Compile warning when building new symmetric key code with openssl 0.9.8
	* Bug #101012: Improve recovery from node fetch timeout in java sync
	* Bug #101013: node fetch timeout in java sync implementation should be longer and settable

javasrc/src/org/ccnx/ccn/utils/ccndcontrol.java
/*
 * A CCNx command line utility for managing prefix registrations.
 *
 * Copyright (C) 2009-2012 Palo Alto Research Center, Inc.
 *
 * This work is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License version 2 as published by the
 * Free Software Foundation.
 * This work is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details. You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */

package org.ccnx.ccn.utils;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.Vector;
import java.util.logging.Level;

import org.ccnx.ccn.CCNHandle;
import org.ccnx.ccn.config.ConfigurationException;
import org.ccnx.ccn.impl.CCNNetworkManager;
import org.ccnx.ccn.impl.CCNNetworkManager.NetworkProtocol;
import org.ccnx.ccn.impl.support.Log;
import org.ccnx.ccn.profiles.ccnd.CCNDaemonException;
import org.ccnx.ccn.profiles.ccnd.FaceManager;
import org.ccnx.ccn.profiles.ccnd.PrefixRegistrationManager;




/**
 * Java utility to
 *
 */
public class ccndcontrol {

	public static String _extraUsage = "";

	public enum Command {Add, Delete};

	public static class RegEntry {
		public Command command;
		public String uri;
		public NetworkProtocol protocol;
		public String host;
		public String hostName;
		public Integer port;
		public Integer flags;
		public Integer multicastTTL;
		public String multicastInterface;

		public Integer faceID;

		public RegEntry() {}
	}

	private static Vector<RegEntry> regList = new Vector<RegEntry>(5);
	private static boolean verbose = false;

	private static void parseString(String in) {
		String [] tokens = in.split("\\s");
		parseFromTokens(tokens, 0, tokens.length);
	}

	private static void parseFromTokens(String [] tokens, int start, int nTokens) {
		RegEntry entry = new ccndcontrol.RegEntry();
		if (nTokens < 1) {
			usage(_extraUsage);
			System.exit(1);
		}
		String tmp = tokens[start];
		if (null == tmp || tmp.length() == 0) {
			System.err.println("Command token either null or of 0 length.");
			usage(_extraUsage);
			System.exit(1);
		}
		/******** add command **********/
		if (tmp.equalsIgnoreCase("add")) {
			entry.command = Command.Add;
			if (nTokens < 3) {
				usage(_extraUsage);
				System.exit(1);
			}
			tmp = tokens[start + 1];
			if (null == tmp || tmp.length() == 0) {
				System.err.println("URI token either null or of 0 length.");
				usage(_extraUsage);
				System.exit(1);
			}
			entry.uri = tokens[start + 1];

			tmp = tokens[start + 2];
			if (null == tmp || tmp.length() == 0) {
				System.err.println("Protocol token either null or of 0 length.");
				usage(_extraUsage);
				System.exit(1);
			}
			entry.protocol = null;
			for (NetworkProtocol p : NetworkProtocol.values()) {
				String pAsString = p.toString();
				if (tmp.equalsIgnoreCase(pAsString)) {
					entry.protocol = p;
					break;
				}
			}
			if (null == entry.protocol) {
				System.err.println("Protocol (" + tmp + ") not valid.");
				usage(_extraUsage);
				System.exit(1);
			}

			tmp = tokens[start + 3];
			if (null == tmp || tmp.length() == 0) {
				System.err.println("Host name either null or of 0 length.");
				usage(_extraUsage);
				System.exit(1);
			}
			String hostNameNumeric = null;
			try {
				InetAddress ipAddr = InetAddress.getByName(tmp);
				hostNameNumeric = ipAddr.getHostAddress();
			} catch (UnknownHostException e) {
				String reason = e.getMessage();
				System.err.println("Host name (" + tmp + ") not found.  reason: " + reason);
				usage(_extraUsage);
				System.exit(1);
			}
			entry.hostName = tmp;
			entry.host = hostNameNumeric;

			if (start + 4 < nTokens && null != tokens[start + 4]) {
				try {
					entry.port = Integer.valueOf(tokens[start + 4]);
				} catch (NumberFormatException e) {
					System.err.println("Port (" + tokens[start + 4] + ") not valid.");
					usage(_extraUsage);
					System.exit(1);
				}
			} else {
				entry.port = CCNNetworkManager.DEFAULT_AGENT_PORT;
			}
			if (start + 5 < nTokens && null != tokens[start + 5]) {
				try {
					entry.flags = Integer.valueOf(tokens[start + 5]);
				} catch (NumberFormatException e) {
					System.err.println("Flags (" + tokens[start + 5] + ") not valid.");
					usage(_extraUsage);
					System.exit(1);
				}
			}
			if (start + 6 < nTokens && null != tokens[start + 6]) {
				try {
					entry.multicastTTL = Integer.valueOf(tokens[start + 6]);
				} catch (NumberFormatException e) {
					System.err.println("Multicast TTL (" + tokens[start + 6] + ") not valid.");
					usage(_extraUsage);
					System.exit(1);
				}
			}
			if (7 < nTokens && null != tokens[start + 7]) {
				entry.multicastInterface = tokens[start + 7];
			}

		} /* add */

		/******** delete command **********/
		else if ((tmp.equalsIgnoreCase("delete") || tmp.equalsIgnoreCase("del"))) {
			entry.command = Command.Delete;
			if (nTokens < 1) {
				System.err.println("Exiting because nTokens < 1 and command is " + tmp + "nTokens: " + nTokens);
				usage(_extraUsage);
				System.exit(1);
			}
			tmp = tokens[start + 1];
			if (null == tmp || tmp.length() == 0) {
				System.err.println("FaceID token either null or of 0 length.");
				usage(_extraUsage);
				System.exit(1);
			}
			try {
				entry.faceID = Integer.valueOf(tmp);
			} catch (NumberFormatException e) {
				System.err.println("Face ID (" + tokens[start + 1] + ") not valid.");
				usage(_extraUsage);
				System.exit(1);
			}
		} /* del */

		else {
			System.err.println("Command (" + tmp + ") not valid.");
			usage(_extraUsage);
			System.exit(1);
		}

		regList.add(entry);
	}

	private static void processConfigFile(String configFile) {
		Log.info("Processing configuration file " + configFile + ".");
		File file = new File(configFile);
		FileReader reader = null;
		try {
			reader = new FileReader(file);
		} catch (FileNotFoundException e) {
			Log.severe("Unable to find " + configFile + ".");
			System.err.println("Unable to find " + configFile + ".");
			System.exit(1);
		}
		BufferedReader read = new BufferedReader(reader);
		String line = null;
		try {
			while ((line = read.readLine()) != null) {
				Log.info("processConfigFile: parsing " + line);
				if (line.charAt(0) == '#') {
					continue;
				}
				parseString(line);
			}
		} catch (IOException e) {
			Log.severe("IO Error (" + e.getMessage() + ") reading " + configFile + ".");
			System.err.println("IO Error (" + e.getMessage() + ") reading " + configFile + ".");
			System.exit(1);
		}
	}



	/**
	 * Function to print out the options for ccndcontrol
	 *
	 * @returns void
	 */
	public static void usage(String extraUsage) {
		/*
		 *     fprintf(stderr,
            "%s [-d] (-f configfile | (add|del) uri proto host [port [flags [mcastttl [mcastif]]]])\n"
            "   -d enter dynamic mode and create FIB entries based on DNS SRV records\n"
            "   -f configfile add or delete FIB entries based on contents of configfile\n"
            "	add|del add or delete FIB entry based on parameters\n",
            progname);

		 */
		System.out.println("usage: ccndcontrol " + extraUsage + "[-v|-vv] add uri protocol host [port [flags [multicastTTL [multicastInterface]]]]\n" +
						    "      ccndcontrol " + extraUsage + "[-v|-vv] del face_id\n" +
						    "      ccndcontrol " + extraUsage + "[-v|-vv] -f configfile");
	}

	/**
	 *Utility function for the ccndcontrol tool.  Initializes the tool,
     * reads the argument list and constructs face to be added.
	 *
	 * @param args Command line arguments:
	 *
	 * @return void
	 */
	@SuppressWarnings("unused")
	public static int executeCommand(String[] args) {
		boolean dynamic = false;
		String configFile = null;
		int startArg = 0;
		Level logLevel = Level.SEVERE;
		verbose = false;

		for (int i = 0; i < args.length; i++) {
			if (i == 0 && args[0].startsWith("[")) {
				_extraUsage = args[0];
				startArg++;
			} else if (args[i].equals("-h")) {
				usage(_extraUsage);
                return(-1);
			} else if (args[i].equals(("-v"))) {
				if (startArg <= i)
					startArg = i + 1;
				verbose = true;
				logLevel = Level.INFO;
			} else if (args[i].equals(("-vv"))) {
				if (startArg <= i)
					startArg = i + 1;
				logLevel = Level.ALL;
			} else if (args[i].equals(("-d"))) {
				if (startArg <= i)
					startArg = i + 1;
				dynamic = true; // never read
			} else if (args[i].equals("-f")) {
				if (args.length < (i + 2)) {
					usage(_extraUsage);
                    return(-1);
				}
				configFile = args[++i];
				if (startArg <= i) {
					startArg = i + 1;
				}
			}
		}
		Log.setDefaultLevel(logLevel);

		if (null == configFile && args.length < startArg + 2) {
			usage(_extraUsage);
            return(-1);
		}

		if (null != configFile) {
			processConfigFile(configFile);
		} else {
			parseFromTokens(args, startArg, args.length);
		}

		int nReg = regList.size();
		for (int i = 0; i < nReg; i++) {
			RegEntry entry = regList.get(i);
			CCNHandle ccnHandle = null;
			FaceManager fHandle = null;
			Integer faceID = null;
			try {
				ccnHandle = CCNHandle.open();
				fHandle = new FaceManager(ccnHandle);
				if (entry.command == Command.Add) {
					faceID = fHandle.createFace(entry.protocol, entry.host, entry.port);
					if (verbose) {
						System.out.println("Created face " + faceID.toString());
					}
					PrefixRegistrationManager pre = new PrefixRegistrationManager(ccnHandle);
					pre.registerPrefix(entry.uri, faceID, entry.flags);
					if (verbose) {
						System.out.println("Added registration for " + entry.uri);
					}

				} else if (entry.command == Command.Delete) {
					fHandle.deleteFace(entry.faceID);
					if (verbose) {
						System.out.println("Deleted face " + entry.faceID.toString() + " in local ccnd");
					}
				} else {
					/* This really can't happen unless the check above was wrong. */
					System.err.println("Internal error.  command (" + entry.command + ") not add or del");
					return(-1);
				}

			} catch (ConfigurationException e) {
				String m = e.getMessage();
				System.err.println(m);
                return(-1);
			} catch (IOException e) {
				String m = e.getMessage();
				System.err.println(m);
                return(-1);
			}catch (CCNDaemonException e) {
				String m = e.getMessage();
				System.err.println(m);
                return(-1);
			} finally {
                if (ccnHandle != null) {
                    ccnHandle.close();
                }
            }
		}

		return(0);
	}

	/**
	 * Main function for the ccndcontrol tool.
	 * Calls executeCommand() to pass args for processing the command.
	 *
	 * @param args Command line arguments:
	 *
	 * @return void
	 */
	public static void main(String[] args) {
        if (executeCommand(args) < 0) {
            System.err.println("Error processing command, unable to complete");
            System.exit(1);
        } else {
            System.out.println("Success");
            System.exit(0);
        }
    }

}
